test/test-runtime-onMessage.js

```diff
@@ -128,70 +130,111 @@ describe("browser-polyfill", () => {
         },
       };
 
-      // Plain value returned.
-      const messageListener = sinon.stub();
-      const firstResponse = "fake reply";
-      // Resolved Promise returned.
-      const secondResponse = Promise.resolve("fake reply 2");
-      // Rejected Promise returned.
-      const thirdResponse = Promise.reject("fake error 3");
+      return setupTestDOMWindow(fakeChrome).then(window => {
+        const listener = sinon.spy((msg, sender, sendResponse) => {
+          sendResponse("Ignored sendReponse callback on returned Promise");
+
+          return Promise.resolve("listener resolved value");
+        });
+
+        const sendResponseSpy = sinon.spy();
+
+        window.browser.runtime.onMessage.addListener(listener);
+
+        ok(fakeChrome.runtime.onMessage.addListener.calledOnce,
+           "runtime.onMessage.addListener should have been called once");
+
+        let wrappedListener = fakeChrome.runtime.onMessage.addListener.firstCall.args[0];
+
+        let returnedValue = wrappedListener("test message", {name: "fake sender"}, sendResponseSpy);
+        equal(returnedValue, true, "the wrapped listener should have returned true");
+
+        ok(listener.calledOnce, "listener has been called once");
 
-      const sendResponseSpy = sinon.spy();
+        // Wait a promise to be resolved and then check the wrapped listener behaviors.
+        return Promise.resolve().then(() => {
+          ok(sendResponseSpy.calledOnce, "sendResponse callback called once");
 
-      messageListener
-        .onFirstCall().returns(firstResponse)
-        .onSecondCall().returns(secondResponse)
-        .onThirdCall().returns(thirdResponse);
+          equal(sendResponseSpy.firstCall.args[0], "listener resolved value",
+                "sendResponse has been called with the expected value");
+        });
+      });
+    });
+
+    it("ignores asynchronous sendResponse calls if the listener does not return true", () => {
+      const fakeChrome = {
+        runtime: {
+          lastError: null,
+          onMessage: {
+            addListener: sinon.spy(),
+          },
+        },
+      };
 
-      let wrappedListener;
+      const waitPromises = [];
 
       return setupTestDOMWindow(fakeChrome).then(window => {
-        window.browser.runtime.onMessage.addListener(messageListener);
+        const listenerReturnsFalse = sinon.spy((msg, sender, sendResponse) => {
+          waitPromises.push(Promise.resolve().then(() => {
+            sendResponse("Ignored sendReponse callback on returned Promise");
```

Comment:
`s/on returned Promise/on returned false/`.

Reply:
Thanks for catching it, it will make easier to read the failure message if it regress here.